packages/next/server/node-polyfill-fetch.js

```diff
@@ -1,23 +1,56 @@
-import fetch, {
-  Headers,
-  Request,
-  Response,
-} from 'next/dist/compiled/node-fetch'
-
 // Polyfill fetch() in the Node.js environment
+
 if (!global.fetch) {
-  const agent = ({ protocol }) =>
-    protocol === 'http:' ? global.__NEXT_HTTP_AGENT : global.__NEXT_HTTPS_AGENT
-  const fetchWithAgent = (url, opts, ...rest) => {
-    if (!opts) {
-      opts = { agent }
-    } else if (!opts.agent) {
-      opts.agent = agent
+  function getFetchImpl() {
+    return global.__NEXT_USE_UNDICI
+      ? require('next/dist/compiled/undici')
+      : require('next/dist/compiled/node-fetch')
+  }
+  // Due to limitation of global configuartion, we have to do this resolution at runtime
+  global.fetch = (...args) => {
+    const fetchImpl = getFetchImpl()
+
+    if (global.__NEXT_USE_UNDICI) {
+      // Undici does not support the `keepAlive` option,
+      // instead we have to pass a custom dispatcher
+      if (
+        !global.__NEXT_HTTP_AGENT_OPTIONS.keepAlive &&
```

Comment:
```suggestion
        !global.__NEXT_HTTP_AGENT_OPTIONS?.keepAlive &&
```

I think this can be undefined, right?

Reply:
:yes: we should probably be typing this global. everything could be marked as `<type> | undefined` too so we handle it properly.